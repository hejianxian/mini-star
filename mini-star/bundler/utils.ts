import fs from 'fs';
import path from 'path';

/**
 * Find plugin dir names
 */
export function getPluginDirs(): string[] {
  const list = fs.readdirSync(path.resolve(process.cwd(), './plugins/'));

  const plugins = list.filter((item) =>
    fs.statSync(path.resolve(process.cwd(), './plugins/', item)).isDirectory()
  );

  return plugins;
}
