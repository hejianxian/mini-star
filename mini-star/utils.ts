/**
 * Create blank ModuleLoader
 * @returns A new ModuleLoader
 */
export function createNewModuleLoader(): ministar.ModuleLoader {
  return {
    resolves: [],
    status: 'new',
    entryFn: null,
    ins: null,
  };
}

/**
 * Set module loader status to `loaded`
 */
export function setModuleLoaderLoaded(
  moduleLoader: ministar.ModuleLoader,
  moduleExport: {} | null
) {
  moduleLoader.status = 'loaded';
  moduleLoader.ins = moduleExport;
  if (Array.isArray(moduleLoader.resolves)) {
    moduleLoader.resolves.forEach((_resolve) => {
      _resolve(moduleExport);
    });
  }
  moduleLoader.resolves = [];
}

function trimDotOfPath(parent: string, self: string) {
  if (!self.startsWith('.')) {
    return self;
  }
  const parentList = parent.split('/');
  const selfList = self.split('/');
  parentList.pop();
  const selfRet: string[] = [];
  selfList.forEach((item) => {
    if (item === '..') {
      parentList.pop();
    }
    if (item !== '..' && item !== '.') {
      selfRet.push(item);
    }
  });
  return parentList.concat(selfRet).join('/');
}

/**
 * A function how to process module path to real path
 */
export function processModulePath(moduleName: string, path: string): string {
  if (path.startsWith('@plugin/')) {
    const name = path.split('/')[1];
    return `${name}/index.js`;
  }

  if (path.endsWith('.js') || path.startsWith('./')) {
    return trimDotOfPath(
      moduleName,
      !path.endsWith('.js') ? `${path}.js` : path
    );
  }

  return path;
}
