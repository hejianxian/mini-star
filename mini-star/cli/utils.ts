import path from 'path';
import { config } from '../bundler/config';

export function getPluginPackagePath(pluginName: string) {
  return path.resolve(
    config.pluginRoot,
    './plugins/',
    pluginName,
    './package.json'
  );
}

/**
 * Return HH:mm
 */
export function getShortTimeStr(): string {
  const now = new Date();

  return `${String(now.getHours()).padStart(2, '0')}:${String(
    now.getMinutes()
  ).padStart(2, '0')}`;
}
