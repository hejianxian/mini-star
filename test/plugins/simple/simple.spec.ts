import { loadHTMLFile, readFiles, setupBuildTest, sleep } from '../../utils';
import path from 'path';

const output = path.join(__dirname, 'dist');

let files: Record<string, string> = {};

beforeAll(() => {
  setupBuildTest(__dirname);

  files = readFiles(output);
});

describe('plugin', () => {
  it('Should build simple plugin', () => {
    expect(Object.keys(files)).toEqual([
      '/bundle.js',
      '/bundle.js.map',
      '/index.html',
      '/plugins/demo/index.js',
      '/plugins/demo/index.js.map',
    ]);
    expect(files['/plugins/demo/index.js']).toMatchSnapshot();
  });
});

describe('runtime', () => {
  it('html', async () => {
    const { dom, logFn } = await loadHTMLFile(
      path.resolve(__dirname, './dist/index.html')
    );

    expect(dom.window.document.body.innerHTML).toMatchSnapshot();

    await sleep(500);

    expect(dom.window.document.body.innerHTML).toMatchSnapshot();

    expect(logFn.mock.calls.length).toBeGreaterThanOrEqual(2);
    expect(logFn.mock.calls).toEqual([['Hello World'], ['Hello Demo!']]);
  });
});
