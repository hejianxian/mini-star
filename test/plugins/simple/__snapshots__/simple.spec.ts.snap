// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin Should build simple plugin 1`] = `
"definePlugin('@plugins/demo', function () { 'use strict';

	console.log(\\"Hello Demo!\\");

});
//# sourceMappingURL=index.js.map
"
`;

exports[`runtime html 1`] = `""`;

exports[`runtime html 2`] = `"<script src=\\"plugins/demo/index.js\\"></script>"`;
